### PR TITLE
Topup module

### PR DESCRIPTION
**Pull request by ssandeep**

### CODE REVIEW

File: proto/topup/v1beta/query.proto [hunk 1]
```diff
@@ -0,0 +1,35 @@
+syntax = "proto3";
+package heimdall.topup.v1beta1;
+
+// import "cosmos/base/query/v1beta1/pagination.proto";
+// this line is used by starport scaffolding # 1
```
Comment (temaniarpit27) [inline proto/topup/v1beta/query.proto:5]: remove this

File: proto/topup/v1beta/query.proto [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    rpc Sequence(QuerySequenceRequest) returns (QuerySequenceResponse) {
+        option (google.api.http).get = "/x/topup/types/msg_topup/{tx_hash}";
+    }
+}
+
+// this line is used by starport scaffolding # 3
+
+message QuerySequenceRequest {
+    option (gogoproto.goproto_getters) = false;
+    string tx_hash = 1;
```
Comment (temaniarpit27) [inline proto/topup/v1beta/query.proto:28]: tx_hash should be converted to bytes right @murthyvitwit

File: x/topup/client/cli/query.go [hunk 3]
```diff
@@ -1,15 +1,14 @@
...
 	"fmt"
-	// "strings"
-
-	"github.com/spf13/cobra"
 
 	"github.com/cosmos/cosmos-sdk/client"
-	// "github.com/cosmos/cosmos-sdk/client/flags"
-	// sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
```
Comment (temaniarpit27) [inline x/topup/client/cli/query.go:9]: dont use viper

File: x/topup/client/cli/query.go [hunk 4]
```diff
@@ -1,15 +1,14 @@
...
 
+	// hmClient "github.com/maticnetwork/heimdall/client"
```
Comment (temaniarpit27) [inline x/topup/client/cli/query.go:11]: remove unnecessary comments

File: x/topup/client/cli/query.go [hunk 5]
```diff
@@ -24,7 +23,53 @@ func GetQueryCmd(queryRoute string) *cobra.Command {
 		RunE:                       client.ValidateCmd,
 	}
 
+	cmd.AddCommand(
+		GetSequenceCmd(),
+	)
 	// this line is used by starport scaffolding # 1
```
Comment (temaniarpit27) [inline x/topup/client/cli/query.go:29]: same

File: x/topup/client/rest/rest.go [hunk 6]
```diff
@@ -15,3 +15,11 @@ const (
 func RegisterRoutes(clientCtx client.Context, r *mux.Router) {
 	// this line is used by starport scaffolding # 2
 }
+
+// func registerQueryRoutes(clientCtx client.Context, r *mux.Router) {
```
Comment (temaniarpit27) [inline x/topup/client/rest/rest.go:19]: remove this comments

File: x/topup/genesis.go [hunk 7]
```diff
@@ -1,17 +1,17 @@
...
+	"github.com/maticnetwork/heimdall/x/topup/types"
+)
 
-// // InitGenesis initializes the capability module's state from a provided genesis
-// // state.
-// func InitGenesis(ctx sdk.Context, k keeper.Keeper, genState types.GenesisState) {
-// }
+// InitGenesis initializes the capability module's state from a provided genesis
+// state.
+func InitGenesis(ctx sdk.Context, k keeper.Keeper, genState types.GenesisState) {
```
Comment (temaniarpit27) [inline x/topup/genesis.go:11]: init genesis empty?

File: x/topup/module.go [hunk 8]
```diff
@@ -36,6 +38,12 @@ type AppModuleBasic struct {
 	cdc codec.Marshaler
 }
 
+// RegisterGRPCGatewayRoutes registers the gRPC Gateway routes for the topup module.
+func (AppModuleBasic) RegisterGRPCGatewayRoutes(clientCtx client.Context, mux *runtime.ServeMux) {
+	// TODO : uncomment below line
```
Comment (temaniarpit27) [inline x/topup/module.go:43]: todo?

File: x/topup/types/codec.go [hunk 9]
```diff
@@ -2,19 +2,19 @@ package types
 
 import (
 	"github.com/cosmos/cosmos-sdk/codec"
-	cdctypes "github.com/cosmos/cosmos-sdk/codec/types"
+    cdctypes "github.com/cosmos/cosmos-sdk/codec/types"
 	// this line is used by starport scaffolding # 1
 )
 
 func RegisterCodec(cdc *codec.LegacyAmino) {
 	// this line is used by starport scaffolding # 2
```
Comment (temaniarpit27) [inline x/topup/types/codec.go:10]: remove unnecessary comments

File: x/topup/types/genesis.go [hunk 10]
```diff
@@ -1,60 +1,15 @@
...
-		if sq == "" {
-			return errors.New("Invalid Sequence")
-		}
-	}
+func (gs GenesisState) Validate() error {
 	return nil
 }
-
-// GetGenesisStateFromAppState returns staking GenesisState given raw application genesis state
-func GetGenesisStateFromAppState(appState map[string]json.RawMessage) GenesisState {
```
Comment (temaniarpit27) [inline x/topup/types/genesis.go:44]: are these ported?

File: x/topup/types/msg.go [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+// Type Implements Msg.
+func (msg MsgTopup) Type() string {
+	return ModuleName
+}
+
+// ValidateBasic Implements Msg.
+func (msg MsgTopup) ValidateBasic() error {
+	// if msg.FromAddress.Empty() {
```
Comment (temaniarpit27) [inline x/topup/types/msg.go:46]: please check this

File: x/topup/types/msg.go [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+// Type Implements Msg.
+func (msg MsgWithdrawFee) Type() string {
+	return "withdraw"
+}
+
+// ValidateBasic Implements Msg.
+func (msg MsgWithdrawFee) ValidateBasic() error {
+	// if msg.UserAddress.Empty() {
```
Comment (temaniarpit27) [inline x/topup/types/msg.go:110]: same

**Review by temaniarpit27 — CHANGES_REQUESTED** [review]

File: proto/topup/v1beta/query.proto (hunk 2 above)
Comment (ssandeep) [reply proto/topup/v1beta/query.proto:28]: No, it was decided that it will be string for now. Murthy has made the changes in staking as well.

**Comment by ssandeep** [comment]
@temaniarpit27 , I have pushed the changes after resolving all the comments. Please review again.

**Review by temaniarpit27 — APPROVED** [review]
lgtm

**Pull request merged**